client_proposal.go

```diff
@@ -0,0 +1,154 @@
+// license that can be found in the LICENSE file.
+
+package prnm
+
+import (
+	"crypto/rand"
+	"math/big"
+
+	"perun.network/go-perun/apps/payment"
+	ethwallet "perun.network/go-perun/backend/ethereum/wallet"
+	"perun.network/go-perun/channel"
+	"perun.network/go-perun/client"
+	"perun.network/go-perun/log"
+	"perun.network/go-perun/wallet"
+)
+
+// ProposeChannel proposes a new channel to the given peer (perunID) and with
+// the given parameters. Returns the newly created channel controller if the
+// channel was successfully created and funded.
+//
+// After the channel got successfully created, the user is required to start the
+// update handler with PaymentChannel.HandleUpdates(UpdateHandler) and to start
+// the channel watcher with PaymentChannel.Watch() on the returned channel
+// controller.
+//
+// It is important that the passed context does not cancel before twice the
+// ChallengeDuration has passed (at least for real blockchain backends with wall
+// time), or the channel cannot be settled if a peer times out funding.
+func (c *Client) ProposeChannel(
+	ctx *Context,
+	perunID *Address,
+	challengeDuration int64,
+	initialBals *BigInts,
+) (*PaymentChannel, error) {
+	alloc := &channel.Allocation{
+		Assets:   []channel.Asset{(*ethwallet.Address)(&assetAddr)},
+		Balances: [][]channel.Bal{initialBals.values},
+	}
+	prop := &client.ChannelProposal{
+		ChallengeDuration: uint64(challengeDuration),
+		Nonce:             nonce(),
+		ParticipantAddr:   c.w.NewAccount().Address(),
+		AppDef:            payment.AppDef(),
+		InitData:          &payment.NoData{},
+		InitBals:          alloc,
+		PeerAddrs:         []wallet.Address{c.onChain.Address(), (*ethwallet.Address)(&perunID.addr)},
+	}
+	_ch, err := c.client.ProposeChannel(ctx.ctx, prop)
+	return &PaymentChannel{_ch}, err
+}
+
+// HandleChannelProposals is the incoming channel proposal handler routine. It
+// must only be started at most once by the user. Incoming channel proposals are
+// handled using the passed handler.
+func (c *Client) HandleChannelProposals(h ProposalHandler) {
+	c.client.HandleChannelProposals(&proposalHandler{c: c, h: h})
+}
+
+type (
+	// A ProposalHandler decides how to handle incoming channel proposals from
+	// other channel network peers.
+	ProposalHandler interface {
+		// Handle is the user callback called by the Client on an incoming channel
+		// proposal.
+		Handle(*ChannelProposal, *ProposalResponder)
+	}
+
+	// proposalHandler implements a client.ProposalHandler wrapping a prnm
+	// ProposalHandler
+	proposalHandler struct {
+		c *Client         // back-reference for ProposalResponder
+		h ProposalHandler // wrapped ProposalHandler
+	}
+
+	// A ChannelProposal describes a proposal to open a new channel.
+	//
+	// The proposer has index 0 and proposee index 1.
+	ChannelProposal struct {
+		PeerPerunID       *Address // The peer proposing the channel.
+		ChallengeDuration int64    // Proposed challenge duration in case of disputes.
```

Comment:
Maybe add that the `ChallengeDuration` is in seconds, i think we dont state that here yet.

Reply:
Good idea, also did it for `ProposeChannel`.